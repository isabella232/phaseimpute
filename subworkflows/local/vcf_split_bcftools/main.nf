include { BCFTOOLS_PLUGINSPLIT  } from '../../../modules/nf-core/bcftools/pluginsplit'
include { BCFTOOLS_QUERY        } from '../../../modules/nf-core/bcftools/query/main'

workflow VCF_SPLIT_BCFTOOLS {
    take:
    ch_vcf          // channel: [ [id, chr, tools], vcf, index ]

    main:

    ch_versions = Channel.empty()

    BCFTOOLS_QUERY(ch_vcf, [], [], []) // List samples

    ch_samples = ch_vcf
        .join(
            BCFTOOLS_QUERY.out.output.splitText().groupTuple()
                .ifEmpty{
                    error "VCF_SPLIT_BCFTOOLS: File does not have any samples information"
                }
        )
        .branch {
            one : it[3].size() == 1
            multiple : it[3].size() > 1
        }

    BCFTOOLS_PLUGINSPLIT(ch_samples.multiple.map{it[0..2]}, [], [], [], [])
    ch_versions = ch_versions.mix(BCFTOOLS_PLUGINSPLIT.out.versions.first())

    ch_vcf_samples = BCFTOOLS_PLUGINSPLIT.out.vcf
        .transpose()
        .map{metaITC, vcf -> [metaITC + [id: vcf.getBaseName().tokenize(".")[0]], vcf]}

    ch_tbi_samples = BCFTOOLS_PLUGINSPLIT.out.tbi
        .transpose()
        .map{metaITC, tbi -> [metaITC + [id: tbi.getBaseName().tokenize(".")[0]], tbi]}

    ch_vcf_tbi_samples = ch_samples.one
        .map{ it[0..2] }
        .mix(ch_vcf_samples
            .join(ch_tbi_samples)
        )

    emit:
    vcf_tbi        = ch_vcf_tbi_samples   // channel: [ [id, chr, tools], vcf, index ]
    versions       = ch_versions          // channel: [ versions.yml ]

}
