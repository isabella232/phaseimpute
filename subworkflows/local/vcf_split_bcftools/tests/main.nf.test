nextflow_workflow {

    name "Test Subworkflow VCF_SPLIT_BCFTOOLS"
    script "../main.nf"

    config "./nextflow.config"

    workflow "VCF_SPLIT_BCFTOOLS"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/vcf_split_bcftools"
    tag "vcf_split_bcftools"

    tag "bcftools"
    tag "bcftools/split"

    test("Split multiple vcf file") {
        setup {
            run("BCFTOOLS_MERGE") {
                script "../../../../modules/nf-core/bcftools/merge/main.nf"
                process {
                """
                input[0] = Channel.of(
                    [
                        [id: "allSamples"],
                        [file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s_imputed.bcf", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s_imputed.bcf", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA20359/NA20359.s_imputed.bcf", checkIfExist:true)],
                        [file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s_imputed.bcf.csi", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s_imputed.bcf.csi", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA20359/NA20359.s_imputed.bcf.csi", checkIfExist:true)],
                        []
                    ]
                )
                input[1] = Channel.of([
                    [id: "GRCh38"],
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz.fai", checkIfExist:true),
                ])
                """
                }
            }
        }
        when {
            workflow {
                """
                input[0] = BCFTOOLS_MERGE.out.vcf.join(BCFTOOLS_MERGE.out.tbi)
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    workflow.out.vcf_tbi.collect{[
                        it[0],
                        path(it[1]).getFileName().toString(),
                        path(it[2]).getFileName().toString()
                    ] },
                    workflow.out.vcf_tbi.collect{
                        path(it[1]).vcf.summary
                    }
                ).match() },
                { workflow.out.vcf_tbi.collect{
                    assert path(it[1]).vcf.sampleCount == 1
                }}
            )
        }
    }

    test("Split one sample vcf file") {
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [id: 'NA12878'],
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s_imputed.bcf", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s_imputed.bcf.csi", checkIfExist:true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    workflow.out.vcf_tbi.collect{[
                        it[0],
                        path(it[1]).getFileName().toString(),
                        path(it[2]).getFileName().toString()
                    ] }
                ).match() }
            )
        }
    }

    test("Split no sample vcf file") {
        tag "error"
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [id: 'dbsnp_146.hg38'],
                    file(params.modules_testdata_base_path + "genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz", checkIfExist:true),
                    file(params.modules_testdata_base_path + "genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz.tbi", checkIfExist:true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.error },
                { assert snapshot(
                    workflow.out.stdout
                ).match()
                }
            )
        }
    }
}
