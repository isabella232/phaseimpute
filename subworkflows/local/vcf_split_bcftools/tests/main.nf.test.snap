{
    "Split one sample vcf file": {
        "content": [
            [

            ],
            [
                [
                    {
                        "id": "NA12878"
                    },
                    "NA12878.s_imputed.bcf",
                    "NA12878.s_imputed.bcf.csi"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-11T16:06:03.822763802"
    },
    "Split no sample vcf file": {
        "content": [
            "VCF_SPLIT_BCFTOOLS: File does not have any samples information"
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-12T14:41:39.506227701"
    },
    "Split multiple vcf file": {
        "content": [
            [
                "versions.yml:md5,6c3351d97e3a99f7a7a3231fc49f92e2"
            ],
            [
                [
                    {
                        "id": "NA12878"
                    },
                    "NA12878.vcf.gz",
                    "NA12878.vcf.gz.tbi"
                ],
                [
                    {
                        "id": "NA19401"
                    },
                    "NA19401.vcf.gz",
                    "NA19401.vcf.gz.tbi"
                ],
                [
                    {
                        "id": "NA20359"
                    },
                    "NA20359.vcf.gz",
                    "NA20359.vcf.gz.tbi"
                ]
            ],
            [
                "VcfFile [chromosomes=[chr21, chr22], sampleCount=1, variantCount=1739, phased=true, phasedAutodetect=true]",
                "VcfFile [chromosomes=[chr21, chr22], sampleCount=1, variantCount=1739, phased=true, phasedAutodetect=true]",
                "VcfFile [chromosomes=[chr21, chr22], sampleCount=1, variantCount=1739, phased=true, phasedAutodetect=true]"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-11T14:42:48.076610625"
    }
}
